cmd/check-markdown-rs/src/main.rs

```diff
@@ -0,0 +1,318 @@
+use comrak::arena_tree::Node;
+use comrak::{
+    nodes::{Ast, NodeValue},
+    parse_document, Arena, ComrakOptions,
+};
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::env;
+use std::fs;
+use url::Url;
+
+// Main function
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let arena = Arena::new();
+    let args: Vec<String> = env::args().collect();
+
+    if args.len() < 2 {
+        eprintln!("Usage: {} <input_file>", args[0]);
+    }
+
+    let input_file_path = env::current_dir()?.join("src").join(args[1].clone());
+    let input = fs::read_to_string(input_file_path)?;
+    let root = parse_document(&arena, &input, &ComrakOptions::default());
+
+    let toc = generate_toc(&root, &arena);
+
+    let link_validation_result = validate_links(&root);
+
+    let structure_validation_result = validate_document_structure(&root);
+
+    let document_statistics = print_document_statistics(&root);
+
+    // Call the generate_output function
+    let document_statistics = print_document_statistics(&root);
+    generate_output(
+        &toc,
+        &link_validation_result,
+        &structure_validation_result,
+        &document_statistics,
+    );
+
+    Ok(())
+}
+
+// Recursive function to print the nodes of the AST
+fn print_nodes<'a>(
+    node: &'a Node<'a, RefCell<Ast>>,
+    arena: &'a Arena<Node<'a, RefCell<Ast>>>,
+    level: usize,
+) {
+    // Create an indent string based on the current recursion level
+    let indent = "  ".repeat(level);
+    // Match the node value to handle different node types
+    match &node.data.borrow().value {
+        // If the node is a Heading node, print its level and text
+        NodeValue::Heading(heading) => println!(
+            "{}Header (level {}): {}",
+            indent,
+            heading.level,
+            get_node_text(&node)
+        ),
```

Comment:
I know these blocks are small, but I'd be inclined to create a set of functions to handle each node type:

- `handle_heading_node(indent, level, text)`
- `handle_text_node(indent, text)`
- `handle_link_node(indent, title, url)`

The advantages of doing this include:

- simplifies `print_nodes()` and makes it easier to read.
- makes it easier to unit test different types of nodes raising confidence in the parts of the code you have tested formally.

Reply:
I think I added what you were looking for.